Add quick search in workspace feature

simark opened the pull request:
This patch adds a "quick" search in workspace feature.  It works using
the "quick open" widget.  The default keybinding to open it is
ctrl-shift-F.  The user types in the widget and the search results
appear as they type.  Selecting a search result opens the file at the
location of the file.

The backend uses ripgrep to implement the search.  There is way to make
ripgrep output a truly unambiguous machine parsable output.  The best
way that people have found is to enable the color output, and use the
ANSI color codes as markers for the fields and matches.

Signed-off-by: Simon Marchi <simon.marchi@ericsson.com>

akosyakov commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+
+@injectable()
+export class QuickSearchInWorkspace implements QuickOpenModel {
+        private currentSearchId: number = -1;
```
I have not looked into changes yet, but formatting looks strange - 8 spaces for indentation?

akosyakov commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -0,0 +1,267 @@
+        }
+
+        /* Search for the string WHAT in directory ROOT.  Return the assigned search id.  */
+        search(what: string, root: string, opts?: SearchInWorkspaceOptions): Promise<number> {
```
Will it be possible to write tests for it?

simark replied:
There is `ripgrep-search-in-workspace-server.spec.ts`.  Or do you mean tests for something else?

akosyakov replied:
sorry, overlooked it, github collapsed it, too many tests :)

akosyakov commented on packages/search-in-workspace/src/common/search-in-workspace-service.ts:
```diff
@@ -0,0 +1,125 @@
+
+import { injectable, inject } from "inversify";
+import { SearchInWorkspaceServer, SearchInWorkspaceClient, SearchInWorkspaceResult, SearchInWorkspaceOptions } from "../common/search-in-workspace-interface";
+import { WorkspaceService } from "@theia/workspace/lib/browser";
```
common cannot use browser, maybe this file should belong to browser?

hexa00 commented on packages/search-in-workspace/src/common/search-in-workspace-service.ts:
```diff
@@ -0,0 +1,125 @@
+
+                        /* Call the client's searchId, but first give it a
+                           chance to record the returned searchId.  */
+                        setTimeout(() => {
```
use process.nextTick ?

simark replied:
This runs in the browser, so it's not available.

hexa00 replied:
Right

hexa00 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -0,0 +1,267 @@
+                                   control characters for colors as guides.  */
+
+                                /* Extract filename (magenta).  */
+                                const filenameRE = /^\x1b\[m\x1b\[35m(.+?)\x1b\[m:/;
```
would be nice to extract the color codes from the string so we can change it easily.

simark replied:
I did this:

```
        private readonly LINE_START = '^\x1b\\[m\x1b\\[32m';
        private readonly LINE_END = '\x1b\\[m:';
        ...
        const lineRE = new RegExp(`${this.LINE_START}(\\d+)${this.LINE_END}`);

```
I think that we the necessary escaping, it makes it a bit more error-prone, but I don't really mind.

hexa00 replied:
I think you could remove some escaping by using ```` ?

simark replied:
I don't think so, `` strings still process backslashes.  So a `\[` would still become a `[` and be wrongly interpreted as the opening of a regex character set.

svenefftinge commented on packages/search-in-workspace/src/browser/search-in-workspace-frontend-module.ts:
```diff
@@ -0,0 +1,46 @@
+export default new ContainerModule(bind => {
+    bind(QuickSearchInWorkspace).toSelf().inSingletonScope();
+
+    /* Command contribution */
```
Could we have comments only if they really add value?

simark replied:
In this case I agree it's not so useful, but in general this is a very subjective topic :)

svenefftinge commented on packages/search-in-workspace/src/common/search-in-workspace-interface.ts:
```diff
@@ -0,0 +1,69 @@
+
+/* Sort search in workspace results according to file, line, character position and then length.  */
+
+export function SearchInWorkspaceResultCompare(a: SearchInWorkspaceResult, b: SearchInWorkspaceResult) {
```
Functions usually start with a lower case letter.

simark replied:
Ok.

svenefftinge commented on packages/search-in-workspace/src/common/search-in-workspace-service.ts:
```diff
@@ -0,0 +1,125 @@
+
+/* Class that will receive the search results from the server.  This is separate
+   from the SearchInWorkspaceService class only to avoid a cycle in the
+   dependency injection.  */
```
Such API comments should be formatted like so:
```
/** 
 * Class that will receive the search results from the server.  This is separate
 * from the SearchInWorkspaceService class only to avoid a cycle in the
 * dependency injection.  
 */
```

simark replied:
Ok.

svenefftinge commented on packages/search-in-workspace/src/common/search-in-workspace-service.ts:
```diff
@@ -0,0 +1,125 @@
+
+        /* All the searches that we have started, that are not done yet (onDone
+           with that searchId has not been called).  */
+        private pendingSearches: Map<number, SearchInWorkspaceCallbacks> = new Map();
```
```
private pendingSearches = new Map<number, SearchInWorkspaceCallbacks>();
```

simark replied:
Ok.

svenefftinge commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+
+        /* If we have an ongoing search, cancel it.  */
+        cancelCurrentSeach() {
+                if (this.currentSearchId >= 0) {
```
Instead of exposing and managing the process numbers, we should use the cancellationToken.

svenefftinge replied:
Actually, I think using the ids is fine. I understand now that with all the async nature and potentially multiple clients, we need to make sure that we can associate matches with an actual search. So please ignore the comment above.

svenefftinge commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+                        return false;
+                }
+
+                /* Search results are 1-based, positions in editors are 0-based.  */
```
we format in-code comments like this
```
// Search results are 1-based, positions in editors are 0-based.
```

simark replied:
Ok.

svenefftinge commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+                };
+
+                /* Show the path relative to the workspace.  */
+                const file = labelProvider.getLongName(new URI('file://' + result.file));
```
we should show the simple file name. The path should be shown after it.
i.e.
**my-functions.ts:30** _src/mypackage_

starting with the path is too much noise before the important segment.
Also, I think the column is not helpful and should be skipped.

simark replied:
Ok, I'll work on that.  Agreed that the column is not really useful.

svenefftinge commented on packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts:
```diff
@@ -0,0 +1,30 @@
+                new JsonRpcConnectionHandler<SearchInWorkspaceClient>
+                        ('/search-in-workspace', client => {
+                                const server = ctx.container.get<SearchInWorkspaceServer>(SearchInWorkspaceServer);
+                                server.setClient(client);
```
This will reset a single client, whenever a new connection is created. I.e. if you have two frontends only the last one will receive results.

simark replied:
What would be the right way to do it?

akosyakov replied:
either each client should receive own instance of `SearchInWorkspaceServer` or use a dispatching client like here: https://github.com/theia-ide/theia/blob/37155c6a897b6ec57bb7765b207dd5389e297011/packages/messages/src/node/messages-backend-module.ts#L18

simark replied:
Ideally, each client would receive its own instance or `SearchInWorkspaceServer`, because searches by one client are not related to searches of another client.  So we don't want the server send `onResult` and `onDone` to other clients than the one that made the request.  I'm not sure how to do that though.

simark replied:
Ah, not making those singletons seems to work:

```diff
diff --git a/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts b/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
index 88908ff..696fe67 100644
--- a/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
+++ b/packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts
@@ -17,7 +17,7 @@ import {
 import { RipgrepSearchInWorkspaceServer } from "./ripgrep-search-in-workspace-server";
 
 export default new ContainerModule(bind => {
-        bind(SearchInWorkspaceServer).to(RipgrepSearchInWorkspaceServer).inSingletonScope();
+        bind(SearchInWorkspaceServer).to(RipgrepSearchInWorkspaceServer);
         bind(ConnectionHandler).toDynamicValue(ctx =>
                 new JsonRpcConnectionHandler<SearchInWorkspaceClient>
                         ('/search-in-workspace', client => {
@@ -26,5 +26,5 @@ export default new ContainerModule(bind => {
                                 client.onDidCloseConnection(() => server.dispose());
                                 return server;
                         })
-        ).inSingletonScope();
+        );
 });

```

akosyakov commented:
Process logging is very verbose. @hexa00 Could we use debug level for an exit with 0 code and log with error level on error or when a process exits with not 0 code?

hexa00 commented:
@akosyakov yes @simark already noted that also and will change it to debug.

Should help in other places too like quick open file.

simark commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+        QuickOpenService,
+        QuickOpenModel,
+        QuickOpenItem,
+        QuickOpenItemOptions
```
I think it gives super long, readable lines and hard to read diffs... but I can change it if you insist.

hexa00 replied:
It's just to be consistent with the rest of theia...

simark replied:
I have opened an issue (#1071) to see if people would be fine with it.

simark commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+@injectable()
+export class QuickSearchInWorkspace implements QuickOpenModel {
+        private currentSearchId: number = -1;
+        private MAX_RESULTS = 100;
```
Done

simark commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,230 @@
+};
+
+@injectable()
+export class SearchInWorkspaceCommandContribution implements CommandContribution {
```
Done

simark commented on packages/search-in-workspace/src/browser/search-in-workspace-frontend-module.ts:
```diff
@@ -0,0 +1,46 @@
+
+    /* The object to call methods on the backend.  */
+    bind(SearchInWorkspaceServer).toDynamicValue(ctx => {
+        const client = ctx.container.get(SearchInWorkspaceClientImpl);
```
I don't understand.  Here we are indeed creating the server, we are just getting the client because we need it to construct the server.

hexa00 replied:
My point is why specify a specific implementation, shouldn't inversify manage that?

simark replied:
I tried to do this change, but I bump into this issue.  The constructor of `SearchInWorkspaceService` calls `client.setService(this);`.  Therefore, it needs to get injected an instance of `SearchInWorkspaceClientImpl` and not `SearchInWorkspaceClient`, because `SearchInWorkspaceClient` does not have that `setService` method.  And I'd rather not add it to the `SearchInWorkspaceClient` interface, it doesn't really make sense there.

simark commented on packages/search-in-workspace/src/common/search-in-workspace-service.ts:
```diff
@@ -0,0 +1,125 @@
+
+@injectable()
+export class SearchInWorkspaceClientImpl implements SearchInWorkspaceClient {
+        service: SearchInWorkspaceClient;
```
No, changed it to private.

simark commented:
New version uploaded, I think I addressed all your comments, I hope I didn't forget any.

akosyakov commented on packages/core/src/browser/label-provider.ts:
```diff
@@ -38,6 +38,10 @@ export interface LabelProviderContribution {
+    /**
+     * returns a directory name for the given element.
+     */
+    getDirectoryName?(element: object): string;
```
could we remove it? instead you should ask for parent uri of you uri and call `getLongName`

akosyakov replied:
label provider is not specific to uris and should not be aware about uri specific concepts

simark replied:
Ah indeed, that makes much more sense.

akosyakov commented on packages/core/src/browser/label-provider.ts:
```diff
@@ -71,6 +75,10 @@ export class DefaultUriLabelProviderContribution implements LabelProviderContrib
+        return uri.path.toString();
+    }
+
+    getDirectoryName(uri: URI): string {
```
please rollback it

akosyakov commented on packages/search-in-workspace/src/browser/quick-search-in-workspace.ts:
```diff
@@ -0,0 +1,229 @@
+                // Show the path relative to the workspace.
+                const uri = new URI('file://' + result.file);
+                const file = labelProvider.getName(uri);
+                const dir = labelProvider.getDirectoryName(uri) + '/';
```
`const dir = labelProvide.getLongName(uri.parent)`

akosyakov commented on packages/search-in-workspace/src/common/search-in-workspace-interface.ts:
```diff
@@ -0,0 +1,93 @@
+ * and then length.
+ */
+
+export function searchInWorkspaceResultCompare(a: SearchInWorkspaceResult, b: SearchInWorkspaceResult) {
```
would be nice to have a namespace for it matching to the interface name:

```ts
export namespace SearchInWorkspaceResult {
    export function compare(...) {...}
}

simark replied:
Is it possible to have both an interface and namespace with the same name?  How do we import both of them?

simark replied:
Nevermind, it just seems to do the right thing.

akosyakov commented on packages/workspace/src/browser/workspace-uri-contribution.ts:
```diff
@@ -87,4 +87,10 @@ export class WorkspaceUriLabelProviderContribution extends DefaultUriLabelProvid
         return short;
     }
+
+    getDirectoryName(element: URI | FileStat): string {
```
should be removed as well

simark commented:
Windows doesn't accept filenames with colons in it, so I changed the tests to only test that on !isWindows.

simark commented:
New version uploaded: fixed tests on Windows (directory separator issue).  For some reason, ripgrep uses two different colors by default for the filename on Windows and Linux.  I added parameters passed to ripgrep so that we choose the colors ourselves.

hexa00 commented:
@simark :) indent is sitll to be fixed ?

simark commented:
> @simark :) indent is sitll to be fixed ?

Where?

simark commented:
New version again, to adapt to the keybindings changes.

akosyakov commented on packages/search-in-workspace/src/node/search-in-workspace-backend-module.ts:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (C) 2017 Erisson and others.
```
2018?

simark replied:
Updated.

simark commented:
Finally, Travis and Appveyor are green.  Any other comments?

hexa00 commented:
@simark reindent like in : quick-search-in-workspace.ts it's still 8 spaces no ?

hexa00 commented:
search-in-workspace-service.ts too, ripgrep-search-in-workspace-server.ts, ripgrep-search-in-workspace-server.ts....

simark commented:
> @simark reindent like in : quick-search-in-workspace.ts it's still 8 spaces no ?

Ahh this indent, you are right.  I thought you were talking about the import indenting, initially.

simark commented:
This is now pushed, if there's anything more to fix, I'll do it in a separate patch.  Thanks for the reviews!
